### PR TITLE
easy_apply

### PR DESCRIPTION
**Pull request by thebigG**
Hello everyone!

These are some of the changes I made:

- "easy_apply" is now a sub-command of JobFunnel.
- easy_apply will look at the master_lis csv file that is passed to JobFunnel and open(by default 1) tab on the user's web browser random job and mark the row that matches that link in the csv with the status  "archive"(by default).
- easy_apply warns the user at 30 or more tabs before opening all of the tabs together on their web browser.
- All the options for easy_apply are prefixed with "ea".
- "--ea_number x" specifies that x number of tabs will opened in the browser and that x number of rows containing those links will be marked with the status the user passed.
- "--ea_output x" specifies that x will be the output file for easy_apply. This output file is used in conjunction with the master_list file. Both the the output_file for easy_apply and master_list will have the same exact data on them after running easy_apply.
- "-ea_status x" says that the rows that are opened on the browser will be marked with status "x". The only valid statuses for now are:["archive", "interview", "offer"].
- As of right now if "easy_apply" is passed to JobFunnel, there is no web scraping done. The main module simply returns if the "easy_apply" option was passed. 

Hope this is clear enough :)

### CODE REVIEW

File: jobfunnel/__main__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env python3
 """Main script.
 
-Scrapes data off several listings, pickles it, and applies search filters.
+Scrapes data off several listings, pickles it, and applies search filters.
```
Comment (PaulMcInnis) [inline jobfunnel/__main__.py]: Can you get rid of the whitespace additions? 

This is best done with an extension or setting for most text editors via “remove training whitespace”

File: jobfunnel/config/parser.py [hunk 2]
```diff
@@ -5,21 +5,39 @@
...
+    easy_apply_parser.add_argument("--ea_number","--number_of_times", help="Number of jobs to apply to",default=1)
+    easy_apply_parser.set_defaults(func=easy_apply.easy_apply, which="easy_apply")
+
+
+def _config_easy_apply_args(config, args):
+    config['ea_output'] = args.ea_output
+    config['ea_status'] = args.ea_status
+    config['ea_number'] = args.ea_number
+    config['easy_apply_func'] = args.func
 def _parse_cli():
```
Comment (PaulMcInnis) [inline jobfunnel/config/parser.py]: Can you add some indentation after _config_easy_apply?

File: jobfunnel/config/parser.py [hunk 3]
```diff
@@ -5,21 +5,39 @@
...
 import os
 import yaml
+from jobfunnel.tools import easy_apply
 
 log_levels = {'critical': logging.CRITICAL, 'error': logging.ERROR,
               'warning': logging.WARNING, 'info': logging.INFO,
               'debug': logging.DEBUG, 'notset': logging.NOTSET}
 
 
+def _parse_easy_apply(parser):
```
Comment (PaulMcInnis) [inline jobfunnel/config/parser.py]: I think these argparse commands make more sense in the existing _parse_cli method

File: jobfunnel/config/parser.py [hunk 4]
```diff
@@ -117,7 +135,8 @@ def _parse_cli():
                         required=False,
                         default=None,
                         help='save duplicates popped by tf_idf filter to file')
-
+    print("parser string:", parser)
```
Comment (PaulMcInnis) [inline jobfunnel/config/parser.py]: Can you remove this print statement?

File: jobfunnel/config/parser.py [hunk 5]
```diff
@@ -135,7 +154,6 @@ def parse_config():
 
     # parse the command line arguments
     cli = _parse_cli()
-
```
Comment (PaulMcInnis) [inline jobfunnel/config/parser.py]: Can you add back these removed indentations?

File: jobfunnel/tools/easy_apply.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        print(f"Are you sure you want to open {number_of_times} tabs on your browser? This could make a low-spec system crash.\n \
+               To go on, type 'y' and ENTER to proceed. Type 'n' to cancel this session.")
+        answer = input()
+        if(answer.lower() != 'y'):
+            return
+    csvfile = open(input_file ,mode='r+')
+    not_new_list  = get_not_new_list(csvfile)
+    csv_row_count = count_csv_rows(csvfile)
+    if(len(not_new_list) == csv_row_count):
+        print("Congrats! You've applied to all the jobs on the list.\
```
Comment (PaulMcInnis) [inline jobfunnel/tools/easy_apply.py]: I think “viewed” makes more sense than applied here - personally i’ll archive jobs i view and dont want to apply to

File: jobfunnel/tools/easy_apply.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    for row in csvreader:
+        if row[0] != "new":
+            not_new.append(count)
+        count += 1
+    csvfile.seek(0)
+    return not_new
+
+def easy_apply(output_file, input_file, status, number_of_times):
+    """
+    This modeule integrates and uses JobFunnel. It is especifically designed to work with JobFunnel.
```
Comment (PaulMcInnis) [inline jobfunnel/tools/easy_apply.py]: Check spelling here.

File: jobfunnel/tools/easy_apply.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    csvfile.seek(0)
+    csvwriter = csv.writer(csvfile, delimiter=',')
+    csvreader = csv.reader(new_csv, delimiter=',')
+    for row in csvreader:
+        csvwriter.writerow(row)
+    print(f"{input_file} has been overwritten with the contents of {output_file}")
+    not_new_count = len(get_not_new_list(csvfile))
+    print(f"Job Application completion:{ ( ((not_new_count)/csv_row_count)*100):.1f}%")
+    csvfile.close()
+    new_csv.close()
```
Comment (PaulMcInnis) [inline jobfunnel/tools/easy_apply.py]: Missing EOF indentation

File: jobfunnel/tools/easy_apply.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    csvfile.seek(0)
+    csvreader = csv.reader(csvfile, delimiter=',')
+    count = 0
+    for row in csvreader:
+        count += 1
+    csvfile.seek(0)
+    print("count=", count)
+    return count
+
+def get_not_new_list(csvfile):
```
Comment (PaulMcInnis) [inline jobfunnel/tools/easy_apply.py]: Missing docstring here

File: jobfunnel/tools/easy_apply.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    """
+    This modeule integrates and uses JobFunnel. It is especifically designed to work with JobFunnel.
+    It will randomly open up as many tabs as you tell it to on your browser from your master_list file that
+    is generated by JobFunnel. It will also mark each one of the randomly selected rows with a status you give it.
+    The default status is 'archive'. After running the easy_apply function, you should see a "updated_master_list.csv" file
+    on the same directory you ran the script on. And your master_list.csv file should have some updated rows with
+    the status you passed.
+    Hope it helps you on your job hunting as much as it has helped me :). 
+    """
+    print("running jobfunnel_easy_apply:",[output_file,input_file,status,number_of_times])
```
Comment (PaulMcInnis) [inline jobfunnel/tools/easy_apply.py]: Super helpful method - does it serve the iser jobs which have already been archived or rejected?

File: jobfunnel/tools/easy_apply.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        print("Congrats! You've applied to all the jobs on the list.\
+               Run JobFunnel again to get more jobs on your master list")
+        return
+    row_nums = []
+    for i in range(number_of_times):
+        row_num  = random.randint(1, csv_row_count)
+        while((row_num in row_nums) or (row_num in not_new_list)):
+            row_num  = random.randint(1, csv_row_count)
+        row_nums.append(row_num)
+    csvfile.seek(0)
```
Comment (PaulMcInnis) [inline jobfunnel/tools/easy_apply.py]: This method should have more in-line documentation

File: readme.md [hunk 12]
```diff
@@ -103,4 +104,47 @@ __*Note*__: `rejected` jobs will be filtered out and will disappear from the out
...
+It will open as many browser tabs as you tell it to with randomly selected job links. It also marks those same jobs with a custom status of your choosing(the default is "archive")
+
+Once you have your scraped "master_list.csv" file, you can start using easy_apply.
+
+	cd JobFunnel/jobfunnel
+
+
+**Make sure that the "data" and "search" folders which contains the master_list are on this directory. Copy them to this directory if you have to. Make sure that there is a "settings.yaml" in "JobFunnel/jobfunnel" as well.**
+
+	./__main__.py easy_apply --ea_number 2
```
Comment (PaulMcInnis) [inline readme.md]: Users shouldn't run main.py directly - this is the same as running “funnel easy apply...”

**Comment by studentbrad** [comment]
I believe that maybe I should play devil's advocate here.
This seems beyond the scope of this project and may be the perfect opportunity to exercise the usefulness of forks in GitHub.
Forks can have features that the repository does not have.  Hence, the author can add features that they think are useful.
Adding this as a feature to the project could ultimately discourage users.
Given how complex the project has already become I am afraid that going down this road could put it in a grave.
I will leave the decision to @PaulMcInnis as he is the owner of this repository.

**Comment by itSeez** [comment]
Recall the [Unix philosophy](https://en.wikipedia.org/wiki/Unix_philosophy). Tools should have one job and do it well.

File: jobfunnel/tools/easy_apply.py (hunk 10 above)
Comment (thebigG) [reply jobfunnel/tools/easy_apply.py]: As of the moment, if something is already "archive" then it just ignores those jobs. The only jobs that are not ignored by easy_apply are those with the status of "new". Do you think this is the best approach, or should I do something about a job that is already "archive"? My thinking was that if something is "archive", then JobFunnel will ignore those jobs. Maybe I misread the docs, but doesn't JobFunnel remove the "archive" jobs? The workflow I had in mind with easy_apply was the following:

1. Scrape jobs with JobFunnel
2. Go get coffee/tea while you wait
3.Use easy_apply to apply for jobs
4. Once you're done applying to all jobs(at this point in my intended use case 100% of jobs are "archive"). 
5. Got back to step 1.

**Comment by PaulMcInnis** [comment]
RE: behaviour with archive - it removes them and adds them to be filtered away next time data comes in. 

I see lots of value in this tool - but I’m gonna side with the community on this one since it goes a fair bit beyond the original intent. 

Please don't be discouraged from contributing in the future however :)

**Pull request closed without merging**